torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,523 @@
+  // (merging) producer and consumer would create a cycle.
+  //
+  // Search for a path from producer to consumer while ignoring the
+  // producer -> consumer edge.
+  size_t upper_bound = consumer->ord;
+  for (auto* child : producer->out_edges()) {
+    if (child == consumer) continue;
+    if (child->visited()) continue; // already visited by dfs
+    if (dfsSearch(DFSDirection::forward, child, consumer, upper_bound, visited)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+
+static bool is_within_bound(DFSDirection direction, size_t value, size_t bound) {
+  if (direction == DFSDirection::forward) {
+    return value < bound; // upper bound
+  } else {
+    return value > bound; // lower bound
+  }
+}
+
+// Searches for a path from start to end via a forward or backward dfs.
+// Returns if a path exists from start to end.
+// In addition, dfsSearch marks vertices as visited and inserts them into visited.
+template <typename T>
+bool DynamicDAG<T>::dfsSearch(
+    DFSDirection direction,
+    Vertex<T>* start,
+    Vertex<T>* end,
+    size_t bound,
+    visited_list<T>& visited) {
+  vertex_list<T> stack;
+
+  auto visit = [&](Vertex<T>* v) {
+    visited.push_back(v);
+    stack.push_back(v);
+  };
+
+  visit(start);
+
+  while (!stack.empty()) {
+    auto* vertex = stack.back();
+    stack.pop_back();
+
+    vertex_list<T>& next_edges = (direction == DFSDirection::forward) ?
+      vertex->out_edges() :
+      vertex->in_edges();
+
+    for (Vertex<T>* next : next_edges) {
+      if (next == end) {
+        // Path found from start to end.
+        visit(next);
+        return true;
+      }
+      if (!next->visited() && is_within_bound(direction, next->ord, bound)) {
+        visit(next);
+      }
+    }
+  }
+  return false;
+}
+
+
+// Reorder deltaB vertices to occur before deltaF vertices.
+template <typename T>
+void DynamicDAG<T>::reorder(vertex_list<T>& deltaF, vertex_list<T>& deltaB) {
+  sort(deltaB);
+  sort(deltaF);
+
+  size_t num_affected = deltaB.size() + deltaF.size();
+
+  // Gather vertices in the desired order. They don't have correct ords yet.
+  std::vector<unique_vertex<T>> desired_vertex_ordering;
+  desired_vertex_ordering.reserve(num_affected);
+  for (auto it = deltaB.begin(); it != deltaB.end(); ++it) {
+    desired_vertex_ordering.push_back(std::move(vertices_.at((*it)->ord)));
+  }
+  for (auto it = deltaF.begin(); it != deltaF.end(); ++it) {
+    desired_vertex_ordering.push_back(std::move(vertices_.at((*it)->ord)));
+  }
+
+  // Sort the ords by merging two already sorted lists into a large sorted list.
+  // input (example): deltaB = { v(1), v(4), v(7) } , deltaF = { v(0), v(2), v(5) }.
+  // output: { 0, 1, 2, 5, 7 }.
+  std::vector<size_t> sorted_ords;
+  sorted_ords.reserve(num_affected);
+
+  // Use deltaB to hold the concatenation of deltaB and deltaF.
+  auto middle = deltaB.size();
+  deltaB.insert(deltaB.end(), deltaF.begin(), deltaF.end());
```

Comment:
Ugh mutating them like that is not perfect since you're taking in references. Just make the arguments values and move them into this function

Reply:
Alternatively, push ords of deltaB and deltaF into sorted_ords and do `inplace_merge` on it!

Reply:
sounds good!